### ISSUE TITLE
Set int dtype of hashing

### ISSUE THREAD

**Issue by thomasaarholt** [pola-rs/polars#4111]
I'm currently using hashing in order to int-encode categorical variables (strings) in my dataframe to use in gradient boosting models. In one column I have a few million unique strings, and `uint32` is an appropriate hash-encoding for this case. In another case I have a few simple strings, like `1->2` up to `8->8` (64 possible combinations). Here it would be nice if I could limit the hashed datatype to `uint8` so as to save memory. And optional argument for dtype could be passed to the python method:

`pl.col("a").hash(dtype=dtype)` with default `dtype=pl.UInt32`, would be great.

**Comment by thomasaarholt** [comment]
I will add that it's currently a bit difficult to use sklearn-type encoding algorithms with polars:

```python
from sklearn.preprocessing import OrdinalEncoder
import polars as pl
import numpy as np

df = pl.DataFrame({"to_encode": ["a", "b", None], "some_value": [1, 2, 3]})

enc = OrdinalEncoder(dtype=np.int8)
df_encode = df.select("to_encode")
data = df_encode.to_numpy()
transformed = enc.fit_transform(data)

df_transformed = pl.DataFrame(
    {col + "2": transformed[:, i] for i, col in enumerate(df_encode.columns)}
)
df.hstack(df_transformed)


shape: (3, 3)
┌───────────┬────────────┬────────────┐
│ to_encode ┆ some_value ┆ to_encode2 │
│ ---       ┆ ---        ┆ ---        │
│ str       ┆ i64        ┆ i8         │
╞═══════════╪════════════╪════════════╡
│ a         ┆ 1          ┆ 0          │
├╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌┤
│ b         ┆ 2          ┆ 1          │
├╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌┤
│ null      ┆ 3          ┆ 2          │
└───────────┴────────────┴────────────┘

```

**Comment by ghuls** [comment]
You can use,
```python
pl.col(“to_encode").cast(pl.Categorical).to_physical().cast(pl.UInt8)
```

**Comment by ritchie46** [comment]
@ghuls answer is perfect. Our local categoricals are exactly ordinal encodings.

**Comment by thomasaarholt** [comment]
This is super neat! Thank you!

**Comment by thomasaarholt** [comment]
String cache solution, but I have to treat the train and test dataset in the same python session:

```python
with pl.StringCache():

    train2 = train.with_columns(
        [
            pl.col("to_encode").hash().alias("hashed"),
            pl.col("to_encode")
            .cast(pl.Categorical)
            .to_physical()
            .cast(pl.UInt8)
            .alias("encoded"),
        ]
    )

    test2 = test.with_columns(
        [
            pl.col("to_encode").hash().alias("hashed"),
            pl.col("to_encode")
            .cast(pl.Categorical)
            .to_physical()
            .cast(pl.UInt8)
            .alias("encoded"),
        ]
    )

train2.filter(pl.col("to_encode") == "albert").vstack(test2.filter(pl.col("to_encode") == "albert"))


shape: (2, 3)
┌───────────┬─────────────────────┬─────────┐
│ to_encode ┆ hashed              ┆ encoded │
│ ---       ┆ ---                 ┆ ---     │
│ str       ┆ u64                 ┆ u8      │
╞═══════════╪═════════════════════╪═════════╡
│ albert    ┆ 2405933837371676327 ┆ 0       │
├╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┤
│ albert    ┆ 2405933837371676327 ┆ 0       │
└───────────┴─────────────────────┴─────────┘
```

**Comment by ritchie46** [comment]
I think it make sense to allow a seeded `StringCache`.

**Comment by ghuls** [comment]
I think we need something like `unify_dictionaries` in pyarrow, which would allow to "fix" local string caches so both will encode the same string as the same integer:
https://arrow.apache.org/docs/python/generated/pyarrow.ChunkedArray.html?highlight=unify_dictionaries#pyarrow.ChunkedArray.unify_dictionaries

**Comment by ritchie46** [comment]
We actually have that. We use when appending.

**Comment by MikeB2019x** [comment]
Has there been further development on this?  My interest wouldn't be so much the type as the length of the hash.  For example, something like `hashlib/shake256` where I can limit the hash code to a given length eg. `shake_256(<string>).hexdigest(20)`
